### PR TITLE
application.create: Rely on the hostApps for detecting discontinued DTs

### PR DESCRIPTION
**Pull request by nghiant2710**
Connects-to: https://github.com/balena-io/balena-sdk/pull/1139

Change-type: patch
Signed-off-by: Trong Nghia Nguyen <nghiant2710@gmail.com>

### CODE REVIEW

File: balena/models/application.py
Comment (thgreasi) [inline balena/models/application.py]: It seems that we already didn't support DT aliases, in which case I would highly suggest to replace `config.get_device_types()` with a query to the `/v6/device_type` resource filtering on the `slug $or name`.

I'm already working on adding the device type aliases to the model, so we can start using them once that is merged.
See: https://github.com/balena-io/open-balena-api/pull/898

To be clear that's part of the effort of reducing our dependence and dropping the deprecated `/device-types/v1` endpoint.

```suggestion
```

File: balena/models/application.py
Comment (thgreasi) [inline balena/models/application.py]:
```suggestion
            raw_query = "$filter=(slug%20eq%20'{slug}')%20or%20(name%20eq%20'{slug}')&$select=id,name&$expand=is_default_for__application($select=is_archived&$filter=is_host%20eq%20true)".format(slug=device_manifest[0]['slug'])
```
and make that request outside of the `if`
and change the `if` to
```
if not device_type_detail:
    raise exceptions.InvalidDeviceType(device_type)
```

File: balena/models/application.py
Comment (nghiant2710) [reply balena/models/application.py]: @thgreasi, I know we're changing to the new endpoint but since the latest SDK uses `/device-types/v1` (https://github.com/balena-io/balena-sdk/blob/master/lib/models/config.ts#L158) so I still want to keep using it here and will make a change when the JS SDK updated.

File: balena/models/application.py
Comment (nghiant2710) [reply balena/models/application.py]: Updated, the request should be inside the `if ` since there will be exception if `device_manifest` is empty.

File: balena/models/application.py
Comment (thgreasi) [reply balena/models/application.py]: @nghiant2710 the only reason that the JS SDK still uses it is in order to support aliases.
Given that the python SDK doesn't support aliases, I don't see any reason use it just for the shake of using it.
I will leave this up to you though, since we can do this change in a follow up.

File: balena/models/application.py
Comment (thgreasi) [reply balena/models/application.py]: I should have said
```
raw_query = "$filter=(slug%20eq%20'{device_type}')%20or%20(name%20eq%20'{device_type}')&$select=id,name&$expand=is_default_for__application($select=is_archived&$filter=is_host%20eq%20true)".format(device_type=device_type)
```
But w/o removing the `/device-types/v1` check, this doesn't have any effect and the second path of the `or` will never be true.

Feel free to revert to the original version of the query if you end up keeping the slug resolution via the `/device-types/v1` .

**Comment by thgreasi** [comment]
@nghiant2710 feel free to self-certify :+1:

File: balena/models/application.py
Comment (nghiant2710) [reply balena/models/application.py]: @thgreasi, I just realise that the output of `/v6/device_type` endpoint contains less information than current `/device-types/v1` which potentially breaks other methods so it's better to do this change in another PR.

**Comment by nghiant2710** [comment]
@balena-ci I self-certify!

**Review by ghost — APPROVED** [review]

**Pull request merged**
